### PR TITLE
Load tests in buildkite

### PR DESCRIPTION
**Pull request by Sushisource**
Buildkite loadtests

Pt. 1 of #96 

Will add more scenarios after this is merged

### CODE REVIEW

File: packages/test/src/activities/index.ts [hunk 1]
```diff
@@ -5,6 +5,15 @@ import { Connection } from '@temporalio/client';
...
 import { cancellableFetch as cancellableFetchInner } from './cancellable-fetch';
 
+// TODO: Get rid of this by providing client via activity context
+function getTestConnection(): Connection {
+  const serverURLOverride = process.env.TEMPORAL_TESTING_SERVER_URL;
+  if (serverURLOverride !== undefined && serverURLOverride.length > 0) {
+    return new Connection({ address: serverURLOverride });
+  }
+  return new Connection();
+}
```
Comment (bergundy) [inline packages/test/src/activities/index.ts:15]: Not that it's even close to being critical but this would have been sufficient
```suggestion
  const address = process.env.TEMPORAL_TESTING_SERVER_URL || undefined;
  return new Connection({ address });
}
```

File: packages/test/src/perf/all-in-one.ts [hunk 2]
```diff
@@ -35,6 +35,12 @@ async function main() {
   try {
     const starterArgs = argsToForward(starterArgSpec, args);
     const starter = spawn('node', [path.resolve(__dirname, 'starter.js'), ...starterArgs], { shell, stdio: 'inherit' });
+    worker.on('exit', (code) => {
+      // If worker dies unceremoniously then also make sure we can exit and not hang
+      if (code !== 0) {
+        kill(starter, 'SIGINT');
```
Comment (bergundy) [inline packages/test/src/perf/all-in-one.ts:41]:
```suggestion
        await kill(starter, 'SIGINT');
```

Comment (bergundy) [reply packages/test/src/perf/all-in-one.ts:41]: This was bugging me too but never got around to fixing, thanks.

File: packages/worker/src/worker.ts [hunk 3]
```diff
@@ -709,7 +709,11 @@ export class Worker<T extends WorkerSpec = DefaultWorkerSpec> {
               const result = await output.activity.run(output.input);
               const status = result.failed ? 'failed' : result.completed ? 'completed' : 'cancelled';
               span.setAttributes({ status });
-              this.log.debug('Activity resolved', { activityId: output.activity.info.activityId, status });
+              this.log.debug('Activity resolved', {
+                activityId: output.activity.info.activityId,
+                status,
+                result: JSON.stringify(result),
```
Comment (bergundy) [inline packages/worker/src/worker.ts:715]: Please revert

File: package.json [hunk 4]
```diff
@@ -33,12 +34,13 @@
...
   },
   "scripts": {
-    "install": "cd packages/docs && npm ci",
+    "install": "test -n $SKIP_DOCS_INSTALL || (cd packages/docs && npm ci)",
     "rebuild": "npm run clean && npm run build",
     "build": "lerna run --stream build",
     "build.watch": "lerna run --stream build.watch",
     "test": "lerna run --stream test",
     "test.watch": "lerna run --stream test.watch",
+    "ci-load": "node ./packages/test/lib/perf/all-in-one.js --server-address $TEMPORAL_TESTING_SERVER_URL",
```
Comment (bergundy) [inline package.json:43]: Why don't we call this `perf` instead of `load`?
I'd keep this consistent, AFAIC, we can rename either the test files or the script name.

File: package.json [hunk 5]
```diff
@@ -33,12 +34,13 @@
     "typescript": "^4.2.3"
   },
   "scripts": {
-    "install": "cd packages/docs && npm ci",
+    "install": "test -n $SKIP_DOCS_INSTALL || (cd packages/docs && npm ci)",
```
Comment (bergundy) [inline package.json:37]: I'm thinking we might want to remove the docs dependencies out of this completely and only install them on demand.
Just not sure about the local dev experience.

File: package.json [hunk 6]
```diff
@@ -7,6 +7,7 @@
     "@commitlint/cli": "^12.1.4",
     "@commitlint/config-conventional": "^12.1.4",
     "@opentelemetry/exporter-jaeger": "^0.20.0",
+    "@opentelemetry/api": "^0.20.0",
```
Comment (bergundy) [inline package.json:10]: I'm still puzzled as to why this breaks only on buildkite, but if it works, let's leave it.

Comment (bergundy) [reply package.json:10]: Especially puzzling is the fact that `package-lock.json` does not reflect this change.

File: .gitignore [hunk 7]
```diff
@@ -1,3 +1,4 @@
+.idea/
```
Comment (bergundy) [inline .gitignore:1]: Sure we can put this here but I usually would rather not put editor specifics in the repo.
Wouldn't you rather put this in a global gitignore on your computer?

File: .buildkite/pipeline.yml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+steps:
+  - label: 'load-test'
+    agents:
+      queue: 'default'
+      docker: '*'
+    timeout_in_minutes: 15
+    plugins:
+      - docker-compose#v3.0.0:
```
Comment (bergundy) [inline .buildkite/pipeline.yml:8]: `3.0.0`? is this strict versioning? Can we make it `<4` a.k.a `^3.0.0`? I'm asking due to lack of familiarity with buildkite

File: .buildkite/docker/docker-compose.yaml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      driver: none
+    ports:
+      - '7233:7233'
+      - '7234:7234'
+      - '7235:7235'
+      - '7239:7239'
+      - '6933:6933'
+      - '6934:6934'
+      - '6935:6935'
+      - '6939:6939'
```
Comment (bergundy) [inline .buildkite/docker/docker-compose.yaml:23]: Not required for CI.
Instead of duplicating the entire file, why don't you download the original?
```docker-compose -f docker-compose.yml -f server.yml up```

File: .buildkite/docker/build-and-run-load.sh [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+npm --version
+
+npm set progress=false
+export SKIP_DOCS_INSTALL=true
+export TEMPORAL_TESTING_SERVER_URL="temporal:7233"
+
+npm ci
+# Try rebuilding once if it fails
+# TODO: Figure out whatever problem causes "common" package to not get compiled when it should
+npm run build || npm run build
```
Comment (bergundy) [inline .buildkite/docker/build-and-run-load.sh:15]: Wow, this is how you solved the build problem?
Let's keep the hack and track the issue.

File: .buildkite/docker/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,17 @@
+FROM rust:1.54
+
+# INSTALL NODE ============================================================================
+# Unfortunately can't use the friendly lts name here easily, because then it gets hard to add to the
+# path. We'll just have to update this as needed.
+ENV NODE_VERSION="v14.17.5"
+RUN curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.38.0/install.sh | bash
+ENV NVM_DIR=/root/.nvm
+RUN . "$NVM_DIR/nvm.sh" && nvm install ${NODE_VERSION}
+RUN . "$NVM_DIR/nvm.sh" && nvm use ${NODE_VERSION}
```
Comment (bergundy) [inline .buildkite/docker/Dockerfile:10]: Why is this required?

File: .buildkite/docker/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,17 @@
+FROM rust:1.54
+
+# INSTALL NODE ============================================================================
+# Unfortunately can't use the friendly lts name here easily, because then it gets hard to add to the
+# path. We'll just have to update this as needed.
+ENV NODE_VERSION="v14.17.5"
```
Comment (bergundy) [inline .buildkite/docker/Dockerfile:6]: Could have used lts and changed the code below to:

`ENV PATH="$(dirname $(. $NVM_DIR/nvm.sh && nvm use 16 &> /dev/null && which node)):$PATH"`

File: .buildkite/pipeline.yml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+steps:
+  - label: 'load-test'
```
Comment (bergundy) [inline .buildkite/pipeline.yml:2]: Wouldn't you rather have separate build and test steps?

**Comment by bergundy** [comment]
I'd also add some assertion about the number WFs processed per second and perhaps check the CPU usage.
I'd run at least 10000 WFs to get more accurate numbers.

File: .buildkite/docker/Dockerfile (hunk 11 above)
Comment (Sushisource) [reply .buildkite/docker/Dockerfile:10]: It's just the easiest way to get node installed in the docker image

File: .gitignore (hunk 7 above)
Comment (Sushisource) [reply .gitignore:1]: Makes it easier for others as well. There's not a reason at all for it to be checked in, no harm in having this line in the gitignore.

File: .buildkite/docker/docker-compose.yaml (hunk 9 above)
Comment (Sushisource) [reply .buildkite/docker/docker-compose.yaml:23]: What original are you referring to here?

File: .buildkite/pipeline.yml (hunk 8 above)
Comment (Sushisource) [reply .buildkite/pipeline.yml:8]: It appears they must be specified exactly

File: .buildkite/pipeline.yml (hunk 13 above)
Comment (Sushisource) [reply .buildkite/pipeline.yml:2]: I kinda would, but the way this docker-compose plugin works it's just much easier to do it this way

File: .buildkite/docker/Dockerfile (hunk 11 above)
Comment (bergundy) [reply .buildkite/docker/Dockerfile:10]: I was asking about `nvm use`, I don't think it does anything here.

File: .buildkite/docker/docker-compose.yaml (hunk 9 above)
Comment (bergundy) [reply .buildkite/docker/docker-compose.yaml:23]: https://github.com/temporalio/docker-compose/blob/main/docker-compose.yml

File: .buildkite/pipeline.yml (hunk 13 above)
Comment (bergundy) [reply .buildkite/pipeline.yml:2]: It's faster now but once we add more parallel steps the build agents will slow down.

File: .gitignore (hunk 7 above)
Comment (bergundy) [reply .gitignore:1]: I will accept this impurity :)

**Review by bergundy — APPROVED** [review]
This can be merged AFAIC.

Had some unresolved comments but no blockers.
